### ISSUE TITLE
[Feature Request] Session folders

### ISSUE THREAD

**Issue by Iurig** [cs0x7f/cstimer#40]
At the same time that I'd like to have a blank session every time I start solving, it's always nice to be able to see progress over time and big averages that can't be done in one sitting, like ao1000. The best way I can think to solve both issues is to have session folders. For example, lets imagine I have a folder called 3x3, and in it, I start daily sessions in which I do 3x3 solves, it would be nice to, at the end of the day, go back to the 3x3 folder and see all the sessions in that folder merged, but maintain the ability to go back to a specific session and see it's times, as well as solve more. I don't have any relevant programming experience, so I don't know how hard that'd be to implement, just thought it was a cool idea.

**Comment by cs0x7f** [comment]
Can I read the suggestion as, to support some cross-session data analysis? Thus, you can select some of sessions (by name, id, scramble type, etc), and do a whole statistics without merge them.

**Comment by Iurig** [comment]
I think that would make my specific example technically possible, but impractical. That's because having one new session every day would clutter the session list and selecting the relevant sessions would become more and more of a hassle if these sessions can't be separated from others just based on stuff like scramble type. Of course, cross session analysis would be welcome, but it seems to me a folder system would be better.

**Comment by cs0x7f** [comment]
Would it be better to use a label system. Thus you can assign a label to each session, and when do statistics, you can select specific label.

**Comment by Iurig** [comment]
That doesn't solve the cluttering problem tho, right? One session a day can become a huge hassle in an year.

**Comment by cs0x7f** [comment]
I see. I'll firstly try doing cross-session statistics, and then think about how to meet the requirement in a simple way.

**Comment by Iurig** [comment]
Maybe if you implement the label system with a max one one label and make it possible to visually merge the sessions with a given label, it wouldn't be such a tool to implement, and it basically would be a folder system without nested folders.

**Comment by cs0x7f** [comment]
I try adding a simple label system. Then, the statistics tools in the right-bottom toolbox is modified to do simple label statistics.

**Comment by tussosedan** [comment]
The labels are a very nice addition, especially if you keep just one big archive session, and one current session -- so you don't have an overload of many sessions, and can see both separate and combined stats.

However, if you'd like to keep the separate details of every mini-session, like Iuri said it's going to cause cluttering of sessions (I actually used to do this before cstimer had dates, it's not fun).

Is there really a simple way to have the session manager display the labels as the top level, collapsed with overall details, and be able to expand into the sessions?

**Comment by cs0x7f** [comment]
I decide to make things simpler. So I remove the labels, and simply add a button (at the left top of session manager table) to allow user to group all continuous sessions with same scramble type without showing all of these sessions.

**Comment by topppits** [comment]
About this:

>At the same time that I'd like to have a blank session every time I start solving

Could you just re-add a shortcut to clear the current session? Or re-add a button. On facebook, on reddit and in your thread many people are missing this feature. If it's possible to implement my suggestion below maybe you can just change it back and instead of the + bring the x back.

>it's always nice to be able to see progress over time and big averages that can't be done in one sitting, like ao1000

The merge function can do that. 

How about for every session you show 3 buttons in the session manager.
- All - if you click this button you get all the solves ever made in that session and you can just keep solving like it is right now
- archive - all solves are "archived" and you see a clean session, but the solves aren't lost. if you click "All" you get to see everything again and could just keep solving with all, or you click archive and start with a clean session.
- Last session - if you click archive all your solves in the list are moved into the archive and last session gets marked, so i know that right now I only see the solves I did recently and this is not the whole list. As long as you didn't archive the last session you can switch between "All" and "Last session" by clicking those two buttons. 

So when I go to the session manager, I see by an indication if right now I see "All" or just the "Last session". On click on those two I can switch between those and if I want to start a new session, I click archive, all my solves from Last session are moved to all and I have a clean session to do solves.

Now we want to see the stats of all those individual small sessions. Implement a marker in that one big session that indicates that someone clicked archive. Then when you get into implementing new statistics you could just show the whole session and also show statistics for each small session. We even have the date, so choosing which session you want to see could go by date like.

I hope I made myself clear and if the stuff with the marker is possible I think this should be implementable and I think it's way more intuitive. It would remove the clutter, everyone can still start with a clean session. We will have the x button for people who don't like to have their solves saved at all but want to start with a clean session and we have the archive option that would be only 2 clicks away.

I don't think we really need to make every individual session available, no one wants to add more solves to that session. We just want to see the stats of each session.

**Comment by Iurig** [comment]
That doesn't work for me. That's because I have multiple different session types with the same scramble time (for example, normal 3x3 practice, CFOP practice, ZZ practice...). So I would keep doing big sessions instead of new ones every day.

**Comment by tussosedan** [comment]
> I don't think we really need to make every individual session accessible , no one wants to add more solves to those old session. We just want to see the stats of each session.

In general I agree, but sometimes I discover I did solves in the wrong session by mistake, and want to migrate them to the right session.


> That doesn't work for me. That's because I have multiple different session types with the same scramble time (for example, normal 3x3 practice, CFOP practice, ZZ practice...). So I would keep doing big sessions instead of new ones every day.

That was at CS's reply, not at topppits, right?

**Comment by topppits** [comment]
>I discover I did solves in the wrong session by mistake, and want to migrate them to the right session.

Input by typing should solve this for most cases. A click on last ao100/1000, select all solves and copy, switch to the correct session, switch to input by typing, paste, press enter, then delete the solves from the wrong session.

I don't think there's really a good way to let us see all sessions and interact with them without having big clutter. But let's see what CS says.

**Comment by Iurig** [comment]
Yes tussosedan, that was targeted at CS's proposal.

**Comment by cs0x7f** [comment]
The 'X' button is reverted back, and can be replaced with '+' by Options->Statistics->Enable session emptying

**Comment by cs0x7f** [comment]
Actually, I'm not going to make cstimer that there is only one reasonable usage of the session manager.
So I add another way of group: grouped by session name (which is also copied when creating new session).
I think sessions grouped with session name or scramble type is enough for most of usage.

**Comment by Iurig** [comment]
Hmmm, that sounds interesting, how would that work exactly?

**Comment by cs0x7f** [comment]
It works in this way:
If two continuous sessions are the same name or same scramble type, then when you click [-] in name column or scramble column, they will be grouped into a single line.
Furthermore, I add a button to sort all sessions by scramble type. (Should I add a similar button to sort sessions by name?)

**Comment by tussosedan** [comment]
So you have to have e.g. sessions named OH, OH, OH, OH etc. one after the other? It's going to be difficult to use it like that... Maybe if the session manager also displayed the dates range of each session, and when grouped showed the dates range and number of solves total?

**Comment by cs0x7f** [comment]
> So you have to have e.g. sessions named OH, OH, OH, OH etc. one after the other? It's going to be difficult to use it like that... Maybe if the session manager also displayed the dates range of each session, and when grouped showed the dates range and number of solves total?

Yes, the next thing I'm going to do is to display date of each session. But the session manager has already been almost fully filled by session name/solves/avgs/scrambles, etc.

**Comment by tussosedan** [comment]
> > So you have to have e.g. sessions named OH, OH, OH, OH etc. one after the other? It's going to be difficult to use it like that... Maybe if the session manager also displayed the dates range of each session, and when grouped showed the dates range and number of solves total?
> 
> Yes, the next thing I'm going to do is to display date of each session. But the session manager has already been almost fully filled by session name/solves/avgs/scrambles, etc.

Perhaps we can save some space: remove the session ID (is it needed for something?), maybe it's possible to remove the reordering arrows and use drag and drop of the lines to reorder, and perhaps to replace the various operations buttons with a single edit button that will open several options.

**Comment by cs0x7f** [comment]
And cstimer is also supported for mobile devices. Recently I'm thinking about whether to add edit/view mode of the session manager or use a multi-line information display.

**Comment by topppits** [comment]
What do you think about my archive suggestion? Would that be implementable - maybe alongside this feature?

I feel like what you implemented now will clutter the session manager. After a year and 1 or more new sessions per day per event, the list will get ridiculously long.
